youtube_dl/extractor/nhk.py

```diff
@@ -3,57 +3,39 @@
-    # Content available only for a limited period of time. Visit
-    # https://www3.nhk.or.jp/nhkworld/en/ondemand/ for working samples.
-    _TESTS = [{
-        # clip
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/video/9999011/',
-        'md5': '256a1be14f48d960a7e61e2532d95ec3',
-        'info_dict': {
-            'id': 'a95j5iza',
-            'ext': 'mp4',
-            'title': "Dining with the Chef - Chef Saito's Family recipe: MENCHI-KATSU",
-            'description': 'md5:5aee4a9f9d81c26281862382103b0ea5',
-            'timestamp': 1565965194,
-            'upload_date': '20190816',
-        },
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/video/2015173/',
-        'only_matching': True,
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/audio/plugin-20190404-1/',
-        'only_matching': True,
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/fr/ondemand/audio/plugin-20190404-1/',
-        'only_matching': True,
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/audio/j_art-20150903-1/',
-        'only_matching': True,
-    }]
-    _API_URL_TEMPLATE = 'https://api.nhk.or.jp/nhkworld/%sod%slist/v7a/episode/%s/%s/all%s.json'
+class NhkBaseInfoExtractor(InfoExtractor):
+    _API_URL_TEMPLATE = 'https://api.nhk.or.jp/nhkworld/%sod%slist/v7a/%s/%s/%s/all%s.json'
 
-    def _real_extract(self, url):
-        lang, m_type, episode_id = re.match(self._VALID_URL, url).groups()
-        if episode_id.isdigit():
-            episode_id = episode_id[:4] + '-' + episode_id[4:]
+    def _get_clean_field(self, episode, key):
+        return episode.get(key + '_clean') or episode.get(key)
 
-        is_video = m_type == 'video'
-        episode = self._download_json(
+    def _list_episodes(self, lang, is_video, is_episode, m_id):
+        return self._download_json(
             self._API_URL_TEMPLATE % (
                 'v' if is_video else 'r',
-                'clip' if episode_id[:4] == '9999' else 'esd',
-                episode_id, lang, '/all' if is_video else ''),
-            episode_id, query={'apikey': 'EJfK8jdS57GqlupFgAfAAwr573q01y6k'})['data']['episodes'][0]
+                'clip' if m_id[:4] == '9999' else 'esd',
+                'episode' if is_episode else 'program',
+                m_id, lang, '/all' if is_video else ''),
+            m_id, query={'apikey': 'EJfK8jdS57GqlupFgAfAAwr573q01y6k'})['data']['episodes']
+
+    def _parse_episode_json(self, episode, lang, is_video):
         title = episode.get('sub_title_clean') or episode['sub_title']
 
-        def get_clean_field(key):
-            return episode.get(key + '_clean') or episode.get(key)
+        pgm_id = episode.get('pgm_id')
+        pgm_no = episode.get('pgm_no')
+
+        if not (pgm_id and pgm_no):
+            raise ExtractorError('Cannot download episode: %s' % compat_str(episode), expected=True)
```

Comment:
i think my previous comment wasn't fully addressed, why are calling str/compat_str on a `dict` here?
after resolving this you can squash the commit into a single commit.

Reply:
Oh, I didn't realize you wanted no `str`/`compat_str`, I thought you were just asking for me to use `compat_str`. As for why I'm calling it, well typically when in a handle error code path, I become a little more careful about what I do. But you know this code base better than me, and I don't do python in my day to day, so if you don't think it is needed, I'm totally fine getting rid of it.

Reply:
Ok, `compat_str` gone and squashed down to one commit.

Reply:
we don't directly use `str` in the project, instead we use `compat_str` to keep the code compatible with all support version of Python, when you call `str/compat` on a dict it won't give a good result, so, i think it should be ommited.